### ISSUE TITLE
depends_on but for build phase (chicken and egg problem for things like SNS topics)

### ISSUE THREAD

**Issue by gabrieljoelc** [aws/copilot-cli#4773]
I have 1 application and 2 services, one existing and one new one:
Existing service:
```yaml
# web/manifest.yml
name: web
type: Load Balanced Web Service

publish:
  topics:
    - name: mytopic
```
New service:
```yaml
# subs/manifest.yml
name: subs
type: Worker Service

subscribe:
  topics:
    - name: mytopic
      service: web
```

```yaml
# pipelines/myrepo-main/manifest.yml

stages:
  - name: dev
    deployments:
      web:
      subs:
        depends_on:
          - web
```

After merging these changes, the `subs` services fails in the build phase with this error:
```
generate workload subs template against environment dev: SNS topic myapp-dev-web-mytopic does not exist in environment dev
--
460 | Cloudformation stack and config files were not generated. Please check build logs to see if there was a manifest validation error.
```

It dawned on me that the `stages.deployments.depends_on` is for the deployment phase but the `subs` build phase "depends on" the `web`'s deployment phase finishing first.

What have others been doing for these chicken-and-egg issues?

**Comment by bvtujo** [comment]
is there any way you can split this into two commits? for example, push a commit which only merges the `web`-related changes, so the SNS topic is created, then push a second commit through the pipeline which includes the `subs` service.

you could also instead push both services at the same time, but modify the manifest of `subs` to not subscribe to anything at first, then push another commit which adds the subscription to `web`'s `mytopic`. 

if you needed to avoid downtime or losing any messages, you could hide the sns logic in `web` behind a feature flag and only release it once the `subs` service is up and running.

**Comment by gabrieljoelc** [comment]
yes, I can do those things but I'd rather not have to coordinate commits if possible. would a `depends_on` for builds help (feature request)? hmm, I guess it wouldn't because we'd have to be able to build _and_ deploy the topic in `web` before `subs` could build.

I'm new to AWS CodePipelines but is it possible to coordinate two different pipelines? One pipeline that builds and deploys `web` and then, if successful, triggers a second pipeline that builds and deploys `subs`?

**Comment by bvtujo** [comment]
I think what you're asking for is possible but would be difficult to implement with Copilot. The easiest way would be to define a pipeline with [graph batch build stage](https://docs.aws.amazon.com/codebuild/latest/userguide/batch-build.html#batch_build_graph) to build & deploy each service in dependency order. 

Adding build/deploy dependencies would require a re-architecture of our pipelines and be a large feature request for us. Please feel free to +1 this issue to help us prioritize. 

In the meantime, is it possible for you to run `copilot svc deploy`? That way you could as a bootstrapping process quickly deploy one service, then the other, and then when you push code changes through the pipeline you shouldn't have these dependency issues.

**Comment by gabrieljoelc** [comment]
I opened this issue after I had already worked around the problem by running `copilot svc deploy -n my-publisher-service` from my local machine and then triggering the pipeline manually. Bootstrapping the topics from a local machine or coordinating commits felt like a sub-optimal user experience for infra-as–code, especially for something as mundane as pub/sub. I can imagine large systems having hundreds of topics and even more subscribers (multiple subscribers per topic) and having to do this kind of coordination for larger teams would be a lot of friction.

This could be an issue with how static SNS/SQS are to built up compared to something like RabbitMQ consumers. The way that RabbitMQ attaches transient consumers to queues linked to [direct exchange](https://www.rabbitmq.com/tutorials/amqp-concepts.html#exchange-direct) topics just by opening a connection makes building subscribers very low friction

If there's no path forward, then feel free to close.

**Comment by bvtujo** [comment]
This problem might be addressed by #4801, adding a `--no-validate` flag to `svc package` and `svc deploy` to skip validations where the customer knows more than Copilot or the SDK. It would take some thought on how to integrate it into pipelines, but it should be doable. Feel free to +1 that issue to help us prioritize!
